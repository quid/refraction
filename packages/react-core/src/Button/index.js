/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import css from '@emotion/css/macro';
import styled from '@emotion/styled/macro';
import { Link } from 'react-router-dom';
import Color from 'color';
import { withFallback as wf, textStyles } from '@quid/theme';
import Icon from '../Icon';

const OKAY = '#039849';
const WARNING = '#FFCE03';
const HAZARD = '#E61E27';

const HEIGHT = {
  small: 24,
  regular: 32,
};

export type Importance =
  | 'primary'
  | 'secondary'
  | 'okay'
  | 'warning'
  | 'hazard';

export type Size = 'regular' | 'small';

type Props = {
  /** The "color" of the button */
  importance?: Importance,
  /** Make the button background transparent */
  transparent?: boolean,
  /** Button height */
  size?: Size,
  /** What to render inside the button */
  children?: React.Node,
  /** Normal `href` URL, if present, the button will render as anchor (`<a />`) */
  href?: string,
  /** react-router URL, if present, the button will render as `Link` */
  to?: string | Object,
  /** when `true`, the button can't be clicked */
  disabled?: boolean,
  /** type to apply to the button HTMLelement when rendering a "real" button */
  type?: 'button' | 'submit' | 'reset',
  /** onClick callback, called when the element is clicked */
  onClick?: (SyntheticEvent<HTMLElement>) => void,
};

const hoverMod = color =>
  Color(color)
    .darken(0.1)
    .string();
const activeMod = color =>
  Color(color)
    .darken(0.2)
    .string();
const disabledMod = color =>
  Color(color)
    .alpha(0.4)
    .string();

const reset = css`
  border: 0;
  padding: 0;
  background: none;
  cursor: pointer;
  font-size: inherit;
  font-family: inherit;
  text-decoration: none;
  box-sizing: border-box;
  color: inherit;
`;

const Button: React.StatelessFunctionalComponent<Props> & {
  Importance: Importance,
  Size: Size,
} = styled(
  React.forwardRef(
    (
      {
        to,
        href,
        transparent,
        size,
        importance,
        disabled,
        type,
        children,
        onClick,
        ...props
      }: Props,
      ref: React.ElementRef<any>
    ) => {
      let Tag, specificProps;
      if (to && !disabled) {
        Tag = Link;
        specificProps = { to };
      } else if ((href || to) && !disabled) {
        Tag = 'a';
        specificProps = { href };
      } else {
        Tag = 'button';
        specificProps = { disabled, type };
      }

      // FIXME: React doesn't detect when buttons are disabled because
      // of a disabled parent fieldset. With this function we can check it
      // by ourselves and avoid to call `onClick` when it shouldn't.
      // https://github.com/facebook/react/issues/7711
      const isButtonDisabled = button =>
        Tag === 'button' && button.matches(':disabled');

      return (
        <Tag
          {...specificProps}
          {...props}
          onClick={evt =>
            onClick && !isButtonDisabled(evt.currentTarget)
              ? onClick(evt)
              : undefined
          }
          ref={ref}
        >
          {React.Children.map(children, node =>
            ['string', 'number'].includes(typeof node) ? (
              <span>{node}</span>
            ) : (
              node
            )
          )}
        </Tag>
      );
    }
  )
)`
  ${reset};

  ${textStyles('normal', 'bold')};
  display: inline-flex;
  vertical-align: top;
  align-items: center;
  border-radius: 2px;
  height: ${props => HEIGHT[props.size]}px;
  padding-left: 0.769em;
  padding-right: 0.769em;
  transition: padding 0.2s ease-in-out, background 0.2s ease-in-out;
  color: ${wf(props => props.theme.colors.black)};
  border: 1px solid transparent;

  ${wf(props => {
    const variations = {
      primary: props.theme.colors.aqua,
      secondary: props.theme.colors.gray2,
      okay: OKAY,
      warning: WARNING,
      hazard: HAZARD,
    };

    const color = variations[props.importance];

    return css`
      background-color: ${color};
      &:hover,
      &[data-state*='hover'] {
        background-color: ${hoverMod(color)};
      }
      &:active,
      &[data-state*='active'] {
        background-color: ${activeMod(color)};
      }
      ${props.disabled ? '&' : '&:disabled'} {
        &,
        &:hover,
        &:focus,
        &:active {
          background-color: ${disabledMod(color)};
        }
      }
      &:focus:not(:focus-visible) {
        outline: 0;
      }
      &:focus-visible,
      &[data-state*='focus'] {
        outline: 0;
        box-shadow: 0 0 0 0.5px ${props.theme.background}, 0 0 2px 2px ${color};
      }
    `;
  })}

  ${props =>
    props.transparent &&
    css`
      background-color: transparent;
      color: ${props.theme.primary};
    `}


  ${props => (props.disabled ? '&' : '&:disabled')} {
    cursor: default;
    &,
    &:hover,
    &:focus,
    &:active {
      color: ${props => props.theme.disabled};
    }
  }

  ${Icon} {
    display: block;
    line-height: 1;
    font-size: ${props => (props.size === 'small' ? 1 : 1.42)}em;
    margin: 5px -6px;
  }

  ${props =>
    React.Children.count(props.children) > 1 &&
    css`
      ${Icon} {
        display: inline-block;
        position: relative;
        font-size: 1em;
        margin-left: 0;
        bottom: -1px;
        margin-left: 0.35em;
        margin-right: 0;
        &:first-child {
          margin-right: 0.35em;
          margin-left: 0;
        }
        &:last-child {
          margin-left: 0.35em;
          margin-right: 0;
        }
      }
    `}
`;

// $FlowFixMe: https://github.com/facebook/flow/issues/5692
Button.defaultProps = {
  importance: 'secondary',
  size: 'regular',
  type: 'button',
};

// @component
export default Button;
