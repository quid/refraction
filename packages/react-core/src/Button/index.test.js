/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import { mount } from 'enzyme';
import Button from '.';
import { Link } from 'react-router-dom';
import Icon from '../Icon';

jest.mock('react-router-dom');

it('renders the expected markup', () => {
  const wrapper = mount(<Button data-action="foo">Hello, World!</Button>);
  expect(wrapper).toMatchSnapshot();
});

it('renders a Button with link', () => {
  const wrapper = mount(
    <Button data-action="foo" to="/">
      Hello, World!
    </Button>
  );
  expect(wrapper.find(Link)).toBeDefined();
});

it('renders a Button with small size', () => {
  const wrapper = mount(<Button size="small">Hello, World!</Button>);
  expect(wrapper).toMatchSnapshot();
});

it('renders a disabled and transparent Button', () => {
  const wrapper = mount(
    <Button transparent disabled>
      Hello, World!
    </Button>
  );
  expect(wrapper).toMatchSnapshot();
});

it('renders a Button with external link', () => {
  const wrapper = mount(
    <Button data-action="foo" href="https://example.org">
      Hello, World!
    </Button>
  );
  expect(wrapper.find('a[href]').props().href).toBe('https://example.org');
});

it('renders a disabled Button with external link', () => {
  const wrapper = mount(
    <Button data-action="foo" disabled href="https://example.org">
      Hello, World!
    </Button>
  );
  expect(wrapper.find('a[href]')).toHaveLength(0);
});

it('renders a link without href if `to` is given but `disabled={true}`', () => {
  const wrapper = mount(
    <Button data-action="foo" to="/" disabled={true}>
      Hello, World!
    </Button>
  );
  expect(wrapper.find(Link)).toHaveLength(0);
});

it('applies Button classes to child if child is Icon', () => {
  const wrapper = mount(
    <Button data-action="foo">
      <Icon name="arrow" />
    </Button>
  );
  expect(
    wrapper
      .find('i')
      .props()
      .className.split(' ')
  ).toHaveLength(2);
});

it('does not apply Button classes to child if child is not Icon', () => {
  const wrapper = mount(
    <Button data-action="foo">
      <div data-target>Hello, World!</div>
    </Button>
  );
  expect(wrapper.find('[data-target]').props().className).toMatchInlineSnapshot(
    `undefined`
  );
});

it('applies Button classes to child if child is Icon with near text', () => {
  const wrapper = mount(
    <Button>
      <Icon name="arrow" />
      Foobar
    </Button>
  );
  expect(
    wrapper
      .find('i')
      .props()
      .className.split(' ')
  ).toHaveLength(2);
});

it('renders a Button with type=submit', () => {
  const wrapper = mount(<Button type="submit">Foobar</Button>);
  expect(wrapper.find('button').props().type).toBe('submit');
});

it('forwards ref properly', () => {
  const ref = React.createRef();
  mount(<Button ref={ref}>Foobar</Button>);
  expect(ref.current).toBeDefined();
});

it('does not trigger a click callback when the button is disabled by a parent fieldset', () => {
  const onClick = jest.fn();
  const wrapper = mount(
    <fieldset disabled>
      <Button onClick={onClick} />
    </fieldset>
  );

  const btn = wrapper.find('button');

  btn.prop('onClick')({ currentTarget: { matches: () => true } });

  expect(onClick).not.toHaveBeenCalled();
});

// it('renders properly with isGroupChild property', () => {
//   const wrapper = mount(<Button isGroupChild />);
//   expect(wrapper).toMatchSnapshot();
// });
