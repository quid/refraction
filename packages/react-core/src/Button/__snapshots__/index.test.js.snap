// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Button with small size 1`] = `
.emotion-0 {
  border: 0;
  padding: 0;
  background: none;
  cursor: pointer;
  font-size: inherit;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  color: inherit;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-weight: bold;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  height: 24px;
  padding-left: 0.769em;
  padding-right: 0.769em;
  -webkit-transition: padding 0.2s ease-in-out,background 0.2s ease-in-out;
  transition: padding 0.2s ease-in-out,background 0.2s ease-in-out;
  color: #121212;
  border: 1px solid transparent;
  background-color: #C7CCD1;
}

.emotion-0:hover,
.emotion-0[data-state*='hover'] {
  background-color: hsl(210,9.8%,72%);
}

.emotion-0:active,
.emotion-0[data-state*='active'] {
  background-color: hsl(210,9.8%,64%);
}

.emotion-0:disabled,
.emotion-0:disabled:hover,
.emotion-0:disabled:focus,
.emotion-0:disabled:active {
  background-color: rgba(199,204,209,0.4);
}

.emotion-0:focus:not(:focus-visible) {
  outline: 0;
}

.emotion-0:focus-visible,
.emotion-0[data-state*='focus'] {
  outline: 0;
  box-shadow: 0 0 0 0.5px #FFFFFF,0 0 2px 2px #C7CCD1;
}

.emotion-0:disabled {
  cursor: default;
}

.emotion-0 .e1t5eso00 {
  display: block;
  line-height: 1;
  font-size: 1em;
  margin: 5px -6px;
}

<Button
  importance="secondary"
  size="small"
  type="button"
>
  <ForwardRef
    className="emotion-0 emotion-1"
    importance="secondary"
    size="small"
    type="button"
  >
    <button
      className="emotion-0 emotion-1"
      onClick={[Function]}
      type="button"
    >
      <span
        key=".0"
      >
        Hello, World!
      </span>
    </button>
  </ForwardRef>
</Button>
`;

exports[`renders a disabled and transparent Button 1`] = `
.emotion-0 {
  border: 0;
  padding: 0;
  background: none;
  cursor: pointer;
  font-size: inherit;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  color: inherit;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-weight: bold;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  height: 32px;
  padding-left: 0.769em;
  padding-right: 0.769em;
  -webkit-transition: padding 0.2s ease-in-out,background 0.2s ease-in-out;
  transition: padding 0.2s ease-in-out,background 0.2s ease-in-out;
  color: #121212;
  border: 1px solid transparent;
  background-color: #C7CCD1;
  background-color: transparent;
}

.emotion-0:hover,
.emotion-0[data-state*='hover'] {
  background-color: hsl(210,9.8%,72%);
}

.emotion-0:active,
.emotion-0[data-state*='active'] {
  background-color: hsl(210,9.8%,64%);
}

.emotion-0,
.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  background-color: rgba(199,204,209,0.4);
}

.emotion-0:focus:not(:focus-visible) {
  outline: 0;
}

.emotion-0:focus-visible,
.emotion-0[data-state*='focus'] {
  outline: 0;
  box-shadow: 0 0 0 0.5px #FFFFFF,0 0 2px 2px #C7CCD1;
}

.emotion-0 {
  cursor: default;
}

.emotion-0 .e1t5eso00 {
  display: block;
  line-height: 1;
  font-size: 1.42em;
  margin: 5px -6px;
}

<Button
  disabled={true}
  importance="secondary"
  size="regular"
  transparent={true}
  type="button"
>
  <ForwardRef
    className="emotion-0 emotion-1"
    disabled={true}
    importance="secondary"
    size="regular"
    transparent={true}
    type="button"
  >
    <button
      className="emotion-0 emotion-1"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        key=".0"
      >
        Hello, World!
      </span>
    </button>
  </ForwardRef>
</Button>
`;

exports[`renders the expected markup 1`] = `
.emotion-0 {
  border: 0;
  padding: 0;
  background: none;
  cursor: pointer;
  font-size: inherit;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  color: inherit;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-weight: bold;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: top;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 2px;
  height: 32px;
  padding-left: 0.769em;
  padding-right: 0.769em;
  -webkit-transition: padding 0.2s ease-in-out,background 0.2s ease-in-out;
  transition: padding 0.2s ease-in-out,background 0.2s ease-in-out;
  color: #121212;
  border: 1px solid transparent;
  background-color: #C7CCD1;
}

.emotion-0:hover,
.emotion-0[data-state*='hover'] {
  background-color: hsl(210,9.8%,72%);
}

.emotion-0:active,
.emotion-0[data-state*='active'] {
  background-color: hsl(210,9.8%,64%);
}

.emotion-0:disabled,
.emotion-0:disabled:hover,
.emotion-0:disabled:focus,
.emotion-0:disabled:active {
  background-color: rgba(199,204,209,0.4);
}

.emotion-0:focus:not(:focus-visible) {
  outline: 0;
}

.emotion-0:focus-visible,
.emotion-0[data-state*='focus'] {
  outline: 0;
  box-shadow: 0 0 0 0.5px #FFFFFF,0 0 2px 2px #C7CCD1;
}

.emotion-0:disabled {
  cursor: default;
}

.emotion-0 .e1t5eso00 {
  display: block;
  line-height: 1;
  font-size: 1.42em;
  margin: 5px -6px;
}

<Button
  data-action="foo"
  importance="secondary"
  size="regular"
  type="button"
>
  <ForwardRef
    className="emotion-0 emotion-1"
    data-action="foo"
    importance="secondary"
    size="regular"
    type="button"
  >
    <button
      className="emotion-0 emotion-1"
      data-action="foo"
      onClick={[Function]}
      type="button"
    >
      <span
        key=".0"
      >
        Hello, World!
      </span>
    </button>
  </ForwardRef>
</Button>
`;
